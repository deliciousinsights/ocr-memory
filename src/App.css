.memory {
  display: flex;
  flex-wrap: wrap;
  width: 300px;
  margin: auto;
  user-select: none;
}
