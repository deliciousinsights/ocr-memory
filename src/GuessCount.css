.memory > .guesses {
  font-size: 1em;
  width: calc(100% - 0.4em);
  margin: 0.2em;
  text-align: right;
  font-family: Menlo, Monaco, Consolas, Inconsolata, 'Courier New', monospace;
}
