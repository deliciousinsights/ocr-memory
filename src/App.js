import React, { Component } from 'react'
import shuffle from 'lodash.shuffle'

import './App.css'

import Card from './Card'
import GuessCount from './GuessCount'
import HallOfFame, { FAKE_HOF } from './HallOfFame'
import HighScoreInput from './HighScoreInput'

const SIDE = 6
export const SYMBOLS = '😀🎉💖🎩🐶🐱🦄🐬🌍🌛🌞💫🍎🍌🍓🍐🍟🍿'
const VISUAL_PAUSE_MSECS = 750

class App extends Component {
  state = {
    cards: this.generateCards(),
    currentPair: [],
    guesses: 0,
    hallOfFame: null,
    matchedCardIndices: [],
  }

  // Arrow fx for binding
  displayHallOfFame = hallOfFame => {
    this.setState({ hallOfFame })
  }

  generateCards() {
    const result = []
    const size = SIDE * SIDE
    const candidates = shuffle(SYMBOLS)
    while (result.length < size) {
      const card = candidates.pop()
      result.push(card, card)
    }
    return shuffle(result)
  }

  getFeedbackForCard(index) {
    const { currentPair, matchedCardIndices } = this.state
    const indexMatched = matchedCardIndices.includes(index)

    if (currentPair.length < 2) {
      return indexMatched || index === currentPair[0] ? 'visible' : 'hidden'
    }

    if (currentPair.includes(index)) {
      return indexMatched ? 'justMatched' : 'justMismatched'
    }

    return indexMatched ? 'visible' : 'hidden'
  }

  // Arrow fx for binding
  handleCardClick = index => {
    const { currentPair } = this.state

    if (currentPair.length === 2) {
      return
    }

    if (currentPair.length === 0) {
      this.setState({ currentPair: [index] })
      return
    }

    this.handleNewPairClosedBy(index)
  }

  handleNewPairClosedBy(index) {
    const { cards, currentPair, guesses, matchedCardIndices } = this.state

    const newPair = [currentPair[0], index]
    const newGuesses = guesses + 1
    const matched = cards[newPair[0]] === cards[newPair[1]]
    this.setState({ currentPair: newPair, guesses: newGuesses })
    if (matched) {
      this.setState({ matchedCardIndices: [...matchedCardIndices, ...newPair] })
    }
    setTimeout(() => this.setState({ currentPair: [] }), VISUAL_PAUSE_MSECS)
  }

  render() {
    const { cards, guesses, hallOfFame, matchedCardIndices } = this.state
    const won = matchedCardIndices.length === cards.length
    return (
      <div className="memory">
        <GuessCount guesses={guesses} />
        {cards.map((card, index) => (
          <Card
            card={card}
            feedback={this.getFeedbackForCard(index)}
            index={index}
            key={index}
            onClick={this.handleCardClick}
          />
        ))}
        {won &&
          (hallOfFame ? (
            <HallOfFame entries={hallOfFame} />
          ) : (
            <HighScoreInput
              guesses={guesses}
              onStored={this.displayHallOfFame}
            />
          ))}
      </div>
    )
  }
}

export default App
