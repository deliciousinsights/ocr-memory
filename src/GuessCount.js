import PropTypes from 'prop-types'
import React from 'react'

import './GuessCount.css'

const GuessCount = ({ guesses }) => <div className="guesses">{guesses}</div>

GuessCount.propTypes = {
  guesses: PropTypes.number.isRequired,
}

export default GuessCount
