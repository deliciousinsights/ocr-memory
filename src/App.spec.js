import { expect } from 'chai'
import React from 'react'
import { shallow } from 'enzyme'
import sinon from 'sinon'

import App, { SYMBOLS } from './App'
import GuessCount from './GuessCount'

describe('<App/>', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(<App />)
  })

  it('contains a zero-guess counter', () => {
    const wrapper = shallow(<App />)

    expect(wrapper).to.contain(<GuessCount guesses={0} />)
  })

  it('has 36 cards', () => {
    const wrapper = shallow(<App />)
    expect(wrapper.find('Card')).to.have.length(36)
  })

  it('should match its reference snapshot', () => {
    const mock = sinon
      .stub(App.prototype, 'generateCards')
      .returns([...SYMBOLS.repeat(2)])
    try {
      const wrapper = shallow(<App />)

      expect(wrapper).to.matchSnapshot()
    } finally {
      mock.restore()
    }
  })
})
