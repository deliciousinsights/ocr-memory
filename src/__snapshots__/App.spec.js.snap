// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> should match its reference snapshot 1`] = `
<div
  className="memory"
>
  <GuessCount
    guesses={0}
  />
  <Card
    card="😀"
    feedback="hidden"
    index={0}
    key="0"
    onClick={[Function]}
  />
  <Card
    card="🎉"
    feedback="hidden"
    index={1}
    key="1"
    onClick={[Function]}
  />
  <Card
    card="💖"
    feedback="hidden"
    index={2}
    key="2"
    onClick={[Function]}
  />
  <Card
    card="🎩"
    feedback="hidden"
    index={3}
    key="3"
    onClick={[Function]}
  />
  <Card
    card="🐶"
    feedback="hidden"
    index={4}
    key="4"
    onClick={[Function]}
  />
  <Card
    card="🐱"
    feedback="hidden"
    index={5}
    key="5"
    onClick={[Function]}
  />
  <Card
    card="🦄"
    feedback="hidden"
    index={6}
    key="6"
    onClick={[Function]}
  />
  <Card
    card="🐬"
    feedback="hidden"
    index={7}
    key="7"
    onClick={[Function]}
  />
  <Card
    card="🌍"
    feedback="hidden"
    index={8}
    key="8"
    onClick={[Function]}
  />
  <Card
    card="🌛"
    feedback="hidden"
    index={9}
    key="9"
    onClick={[Function]}
  />
  <Card
    card="🌞"
    feedback="hidden"
    index={10}
    key="10"
    onClick={[Function]}
  />
  <Card
    card="💫"
    feedback="hidden"
    index={11}
    key="11"
    onClick={[Function]}
  />
  <Card
    card="🍎"
    feedback="hidden"
    index={12}
    key="12"
    onClick={[Function]}
  />
  <Card
    card="🍌"
    feedback="hidden"
    index={13}
    key="13"
    onClick={[Function]}
  />
  <Card
    card="🍓"
    feedback="hidden"
    index={14}
    key="14"
    onClick={[Function]}
  />
  <Card
    card="🍐"
    feedback="hidden"
    index={15}
    key="15"
    onClick={[Function]}
  />
  <Card
    card="🍟"
    feedback="hidden"
    index={16}
    key="16"
    onClick={[Function]}
  />
  <Card
    card="🍿"
    feedback="hidden"
    index={17}
    key="17"
    onClick={[Function]}
  />
  <Card
    card="😀"
    feedback="hidden"
    index={18}
    key="18"
    onClick={[Function]}
  />
  <Card
    card="🎉"
    feedback="hidden"
    index={19}
    key="19"
    onClick={[Function]}
  />
  <Card
    card="💖"
    feedback="hidden"
    index={20}
    key="20"
    onClick={[Function]}
  />
  <Card
    card="🎩"
    feedback="hidden"
    index={21}
    key="21"
    onClick={[Function]}
  />
  <Card
    card="🐶"
    feedback="hidden"
    index={22}
    key="22"
    onClick={[Function]}
  />
  <Card
    card="🐱"
    feedback="hidden"
    index={23}
    key="23"
    onClick={[Function]}
  />
  <Card
    card="🦄"
    feedback="hidden"
    index={24}
    key="24"
    onClick={[Function]}
  />
  <Card
    card="🐬"
    feedback="hidden"
    index={25}
    key="25"
    onClick={[Function]}
  />
  <Card
    card="🌍"
    feedback="hidden"
    index={26}
    key="26"
    onClick={[Function]}
  />
  <Card
    card="🌛"
    feedback="hidden"
    index={27}
    key="27"
    onClick={[Function]}
  />
  <Card
    card="🌞"
    feedback="hidden"
    index={28}
    key="28"
    onClick={[Function]}
  />
  <Card
    card="💫"
    feedback="hidden"
    index={29}
    key="29"
    onClick={[Function]}
  />
  <Card
    card="🍎"
    feedback="hidden"
    index={30}
    key="30"
    onClick={[Function]}
  />
  <Card
    card="🍌"
    feedback="hidden"
    index={31}
    key="31"
    onClick={[Function]}
  />
  <Card
    card="🍓"
    feedback="hidden"
    index={32}
    key="32"
    onClick={[Function]}
  />
  <Card
    card="🍐"
    feedback="hidden"
    index={33}
    key="33"
    onClick={[Function]}
  />
  <Card
    card="🍟"
    feedback="hidden"
    index={34}
    key="34"
    onClick={[Function]}
  />
  <Card
    card="🍿"
    feedback="hidden"
    index={35}
    key="35"
    onClick={[Function]}
  />
</div>
`;
